File: bci/api/python/wia_bci/core/event_emitter.py
"""WIA BCI Event Emitter."""

from collections import defaultdict
from typing import Any, Callable, Optional

from wia_bci.types.events import EventType

EventHandler = Callable[..., None]


class BciEventEmitter:
    """Type-safe event emitter for BCI events."""

    def __init__(self) -> None:
        self._listeners: dict[EventType, list[EventHandler]] = defaultdict(list)
        self._once_handlers: set[EventHandler] = set()

    def on(self, event: EventType, handler: EventHandler) -> "BciEventEmitter":
        """Subscribe to an event."""
        self._listeners[event].append(handler)
        return self

    def off(self, event: EventType, handler: EventHandler) -> "BciEventEmitter":
        """Unsubscribe from an event."""
        if event in self._listeners:
            try:
                self._listeners[event].remove(handler)
            except ValueError:
                pass
        return self

    def once(self, event: EventType, handler: EventHandler) -> "BciEventEmitter":
        """Subscribe to an event once."""
        self._once_handlers.add(handler)
        self._listeners[event].append(handler)
        return self

    def emit(self, event: EventType, data: Optional[Any] = None) -> "BciEventEmitter":
        """Emit an event."""
        handlers = self._listeners.get(event, []).copy()

        for handler in handlers:
            try:
                if data is None:
                    handler()
                else:
                    handler(data)
            except Exception as e:
                print(f"Error in event handler for '{event}': {e}")

            # Remove once handlers
            if handler in self._once_handlers:
                self._once_handlers.discard(handler)
                try:
                    self._listeners[event].remove(handler)
                except ValueError:
                    pass

        return self

    def remove_all_listeners(self, event: Optional[EventType] = None) -> "BciEventEmitter":
        """Remove all listeners for an event (or all events)."""
        if event is not None:
            self._listeners.pop(event, None)
        else:
            self._listeners.clear()
            self._once_handlers.clear()
        return self

    def listener_count(self, event: EventType) -> int:
        """Get listener count for an event."""
        return len(self._listeners.get(event, []))

    def event_names(self) -> list[EventType]:
        """Get all events with listeners."""
        return list(self._listeners.keys())

File: bci/api/typescript/src/core/EventEmitter.ts
/**
 * WIA BCI Event Emitter
 * @module wia-bci/core/EventEmitter
 */

import type { EventType, EventHandler, EventDataMap } from '../types';

/**
 * Type-safe Event Emitter for BCI events
 */
export class BciEventEmitter {
  private listeners: Map<EventType, Set<EventHandler<EventType>>> = new Map();

  /**
   * Subscribe to an event
   */
  on<T extends EventType>(event: T, handler: EventHandler<T>): this {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)!.add(handler as EventHandler<EventType>);
    return this;
  }

  /**
   * Unsubscribe from an event
   */
  off<T extends EventType>(event: T, handler: EventHandler<T>): this {
    const handlers = this.listeners.get(event);
    if (handlers) {
      handlers.delete(handler as EventHandler<EventType>);
    }
    return this;
  }

  /**
   * Subscribe to an event once
   */
  once<T extends EventType>(event: T, handler: EventHandler<T>): this {
    const onceHandler = ((data: EventDataMap[T]) => {
      this.off(event, onceHandler as EventHandler<T>);
      (handler as (data: EventDataMap[T]) => void)(data);
    }) as EventHandler<T>;

    return this.on(event, onceHandler);
  }

  /**
   * Emit an event
   */
  emit<T extends EventType>(event: T, data?: EventDataMap[T]): this {
    const handlers = this.listeners.get(event);
    if (handlers) {
      handlers.forEach((handler) => {
        try {
          if (data === undefined) {
            (handler as () => void)();
          } else {
            (handler as (d: EventDataMap[T]) => void)(data);
          }
        } catch (error) {
          console.error(`Error in event handler for '${event}':`, error);
        }
      });
    }
    return this;
  }

  /**
   * Remove all listeners for an event (or all events)
   */
  removeAllListeners(event?: EventType): this {
    if (event) {
      this.listeners.delete(event);
    } else {
      this.listeners.clear();
    }
    return this;
  }

  /**
   * Get listener count for an event
   */
  listenerCount(event: EventType): number {
    return this.listeners.get(event)?.size ?? 0;
  }

  /**
   * Get all events with listeners
   */
  eventNames(): EventType[] {
    return Array.from(this.listeners.keys());
  }
}
